pkg/provider/barbican/client.go

```diff
@@ -0,0 +1,268 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package barbican
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/pkg/constants"
+	"github.com/external-secrets/external-secrets/pkg/metrics"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/keymanager/v1/secrets"
+)
+
+const (
+	errBarbicanStore        = "received invalid Barbican resource"
+	errFetchSAKSecret       = "could not fetch SecretAccessKey secret: %w"
+	errMissingSAK           = "missing SecretAccessKey"
+	errKeyNotFound          = "key '%s' does not exist"
+	errKeyAlreadyExists     = "key '%s' already exists"
+	errJSONSecretUnmarshal  = "unable to unmarshal secret: %w"
+	errInvalidStore         = "invalid store"
+	errInvalidStoreSpec     = "invalid store spec"
+	errInvalidStoreProv     = "invalid store provider"
+	errInvalidBarbicanProv  = "invalid barbican secrets manager provider"
+	errInvalidAuthSecretRef = "invalid auth secret ref: %w"
+)
+
+type config struct {
+	// OpenStack Auth Url
+	AuthUrl string `json:"auth_url"`
+
+	// The Domain of the user.
+	UserDomain string `json:"user_domain"`
+
+	// The user name. If you do not provide a user name and password, you must provide a token.
+	Username string `json:"username"`
+
+	// The password for the user.
+	Password string `json:"password"`
+
+	// The project name. Both the Project ID and Project Name are optional.
+	ProjectName string `json:"project_name"`
+
+	// ServiceName [optional] is the service name for the client (e.g., "nova") as it
+	// appears in the service catalog. Services can have the same Type but a
+	// different Name, which is why both Type and Name are sometimes needed.
+	ServiceName string `json:"service_name"`
+
+	// Region [required] is the geographic region in which the endpoint resides,
+	// generally specifying which datacenter should house your resources.
+	// Required only for services that span multiple regions.
+	Region string `json:"region"`
+}
```

Comment:
Instead of reading and unmarshalling the secret into this config struct, can we move this to the SecretStore CRD and only pull the Password from a secret? This seems to be the only secret value. 
This would improve UX so users don't have to serialise their credentials into a JSON and store it as a Kind=Secret.

Reply:
hey @artashesbalabekyan are you still working on this? If you need more time i can convert it into a draft or close it if you're low on time

Reply:
Hi @moolen . I understood your approach. I agree. We can store all the data in SecretStore CRD and pull only the password from a Secret. I will make a fix

Reply:
> Instead of reading and unmarshalling the secret into this config struct, can we move this to the SecretStore CRD and only pull the Password from a secret? This seems to be the only secret value. This would improve UX so users don't have to serialise their credentials into a JSON and store it as a Kind=Secret.

@moolen I did it. You can check it.